pkg/api/norman/store/cluster/cluster_store.go

```diff
@@ -306,14 +313,42 @@ func loadDataFromTemplate(clusterTemplateRevision *v3.ClusterTemplateRevision, c
 			return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error processing clusterTemplate answers")
 		}
 		keyParts := strings.Split(question.Variable, ".")
-		values.PutValue(dataFromTemplate, val, keyParts...)
+		if strings.HasPrefix(question.Variable, "rancherKubernetesEngineConfig.privateRegistries") {
+			// for example: question.Variable = rancherKubernetesEngineConfig.privateRegistries[0].url
+			index, err := getIndexFromQuestion(question.Variable)
+			if err != nil {
+				return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error processing clusterTemplate answers to private registry")
+			}
+			key := keyParts[len(keyParts)-1]
+			if _, ok := registryMap[index]; ok {
+				registryMap[index][key] = val
+			} else {
+				registryMap[index] = map[string]interface{}{
+					key: val,
+				}
+			}
+		} else {
+			values.PutValue(dataFromTemplate, val, keyParts...)
+		}
 
 		questionMap, err := convert.EncodeToMap(question)
 		if err != nil {
 			return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error reading clusterTemplate questions")
 		}
 		revisionQuestions = append(revisionQuestions, questionMap)
 	}
+	registries, err := convertRegistryMapToSliceInOrder(registryMap)
+	if err != nil {
+		return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error processing clusterTemplate answers to private registry")
+	}
+	// set the first registry in the list as the default registry
+	for i := 0; i < len(registries); i++ {
+		registries[i][managementv3.PrivateRegistryFieldIsDefault] = false
+	}
+	registries[0][managementv3.PrivateRegistryFieldIsDefault] = true
```

Comment:
Can this logic be added to `convertRegistryMapToSliceInOrder` so we only loop through these items once?

At the very least, the loop variable `i` here should start at 1 instead of 0.

Reply:
at first, I thought the `convertRegistryMapToSliceInOrder` function should not modify the content, but now I think it makes more sense to just let it return the finalized data that can be used for the next step, and of course, saves a loop.